r2-core/src/main/java/com/linkedin/r2/message/RequestContext.java

```diff
@@ -131,4 +140,204 @@ public String toString()
+
+    public AttrsMap() {
+      map = new HashMap<>();
+    }
+
+    public AttrsMap(Map<K, V> other) {
+      map = new HashMap<>(other);
+    }
+
+    /**
+     * Create a clone of another AttrsMap.
+     *
+     * This is more efficient than the generic Map-based constructor for AttrsMap since it synchronizes only once for
+     * the entire copy rather instead of also synchronizing for every element read.
+     *
+     * @param other Another AttrsMap to copy.
+     */
+    public AttrsMap(AttrsMap<K, V> other) {
+      other.lock.lock();
+      try {
+        map = new HashMap<>(other.map);
+      } finally {
+        other.lock.unlock();
+      }
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      lock();
+      try {
+        return map.equals(other);
+      } finally {
+        unlock();
+      }
+    }
+
+    @Override
+    public int hashCode() {
+      lock();
+      try {
+        return map.hashCode();
+      } finally {
+        unlock();
+      }
+    }
+
+    @Override
+    public int size() {
+      lock();
+      try {
+        return map.size();
+      } finally {
+        unlock();
+      }
+    }
+
+    @Override
+    public boolean isEmpty() {
+      lock();
+      try {
+        return map.isEmpty();
+      } finally {
+        unlock();
+      }
+    }
+
+    @Override
+    public boolean containsKey(Object key) {
+      lock();
+      try {
+        return map.containsKey(key);
+      } finally {
+        unlock();
+      }
+    }
+
+    @Override
+    public boolean containsValue(Object value) {
+      lock();
+      try {
+        return map.containsValue(value);
+      } finally {
+        unlock();
+      }
+    }
+
+    @Override
+    public V get(Object key) {
+      lock();
+      try {
+        return map.get(key);
+      } finally {
+        unlock();
+      }
+    }
+
+    @Override
+    public V put(K key, V value) {
+      lock();
+      try {
+        return map.put(key, value);
+      } finally {
+        unlock();
+      }
+    }
+
+    @Override
+    public V remove(Object key) {
+      lock();
+      try {
+        return map.remove(key);
+      } finally {
+        unlock();
+      }
+    }
+
+    @Override
+    public void putAll(Map<? extends K, ? extends V> m) {
+      lock();
+      try {
+        map.putAll(m);
+      } finally {
+        unlock();
+      }
+    }
+
+    @Override
+    public void clear() {
+      lock();
+      try {
+        map.clear();
+      } finally {
+        unlock();
+      }
+    }
+
+    @Override
+    public Set<K> keySet() {
+      lock();
+      try {
+        return map.keySet();
+      } finally {
+        unlock();
+      }
+    }
+
+    @Override
+    public Collection<V> values() {
+      lock();
+      try {
+        return map.values();
+      } finally {
+        unlock();
+      }
+    }
+
+    @Override
+    public Set<Entry<K, V>> entrySet() {
+      lock();
+      try {
+        return map.entrySet();
```

Comment:
This exposes the entries to clients, which are able to make modifications and bypass your locking, leading to continued ConcurrentModificationExceptions. Same for keySet and values.